File: thinking-categorically/applicative-functor/applicative-functor-laws/composition-law.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Program
{
    static void Main()
    {
        // u = [(+1)]
        var u = new List<Func<int, int>> { x => x + 1 };

        // v = [(*2)]
        var v = new List<Func<int, int>> { x => x * 2 };

        // w = [10]
        var w = new List<int> { 10 };

        // Function composition: f ∘ g
        Func<Func<int, int>, Func<int, int>, Func<int, int>> compose =
            (f, g) => x => f(g(x));

        // Left: pure (.) <*> u <*> v <*> w
        var left = new List<Func<Func<int, int>, Func<int, int>, Func<int, int>>> { compose }
            .SelectMany(c => u
            .SelectMany(f => v
            .SelectMany(g => w
            .Select(x => c(f, g)(x))))).ToList();

        // Right: u <*> (v <*> w)
        var right = u
            .SelectMany(f => v
            .SelectMany(g => w
            .Select(x => f(g(x))))).ToList();

        Console.WriteLine($"Left:  {string.Join(", ", left)}");
        Console.WriteLine($"Right: {string.Join(", ", right)}");
        Console.WriteLine($"Equal? {left.SequenceEqual(right)}");
    }
}

File: thinking-categorically/applicative-functor/applicative-functor-laws/composition-law.ts
function main() {
  // u, v are arrays of functions; w is a value array
  const u: Array<(x: number) => number> = [x => x + 1];
  const v: Array<(x: number) => number> = [x => x * 2];
  const w: number[] = [10];

  // Function composition: (f ∘ g)(x) = f(g(x))
  const compose = <A, B, C>(f: (b: B) => C, g: (a: A) => B) => (x: A) => f(g(x));

  // Left: pure (.) <*> u <*> v <*> w
  const left = [compose].flatMap(c =>
    u.flatMap(f =>
      v.flatMap(g =>
        w.map(x => c(f, g)(x))
      )
    )
  );

  // Right: u <*> (v <*> w)
  const right = u.flatMap(f =>
    v.flatMap(g =>
      w.map(x => f(g(x)))
    )
  );

  console.log("Left: ", left);   // [21]
  console.log("Right:", right);  // [21]
  console.log("Equal?", arraysEqual(left, right)); // true
}

function arraysEqual<T>(a: T[], b: T[]): boolean {
  return a.length === b.length && a.every((v, i) => v === b[i]);
}

main();
